src/cargo/ops/lockfile.rs

```diff
@@ -97,6 +91,19 @@ pub fn write_pkg_lockfile(ws: &Workspace, resolve: &Resolve) -> CargoResult<()>
     })
 }
 
+fn are_equal_lockfiles(mut orig: String, current: &str, strict: bool) -> bool {
+    if has_crlf_line_endings(&orig) {
+        orig = orig.replace("\r\n", "\n");
+    }
+
+    // Old lockfiles have unused `[root]` section,
+    // just ignore it if we are in the `--frozen` mode.
+    if !strict && orig.starts_with("[root]") {
+        orig = orig.replacen("[root]", "[[package]]", 1);
```

Comment:
This isn't *quite* right because IIRC the `[root]` was the lexicographically last package (if you have a bunch of path dependencies it's the last one alphabetically). Essentially I've seen `cargo build` on nightly move around this section.

That being said I'm sort of ok just comparing TOML here, I don't think we need to compare literal encodings per se, but rather if Cargo generates one encoding and the encoding on the filesystem is the same, it seems fine to avoid rewriting?

Reply:
Ah, an excellent catch!

I initially though of comparing TOMLs, but then decided that it might be too slow (because of the hyper-optimized `has_crlf`). But we certainly can do anything inside this if, as it's a cold path.

Reply:
Rebased!